### DISCUSSION TITLE
page.rect does not match with the real page size

### DISCUSSION THREAD

**Discussion by jmgomezsoriano** [pymupdf/PyMuPDF#3324] [Q&A]
I've made this code:
```python
doc = fitz.Document(join('examples', 'doc.pdf'), filetype="pdf")
page = doc[0]
images = page.get_images(full=True)
image = images[0]
page.draw_rect(page.rect, color=(0, 0, 1), width=2)
```

But it draws a rectangle smaller than the real page size and not centered in the fist of the page. The page.rect gives me:

Rect(0.0, 0.0, 595.2757568359375, 841.8900146484375)

And trying what is the best rectangle to matches with the page I'll found more or less this:

Rect(0, -279.95, 793.2, 841.4)

Does anyone know the reason of this discrepancy? In other PDF documents it matches, but it doesn't with a PDF which only have a scanned image.

Thanks.

**Comment by JorjMcKie** [comment]
You did not supply your example PDF, so I have to assume a few things.
I am sure you will find the page's rectangle to be defined as an ISO A4 page when you do `print(doc.xref_object(page.xref)`. Something like this:
```python
In [4]: print(doc.xref_object(page.xref))
<<
  /Type /Page
  /Annots [ 10 0 R ]
  /Parent 4 0 R
  /Rotate 0
  /Contents 11 0 R
  /MediaBox [ 0 0 595.32 841.92 ]  # <===
  /Resources <<
    /Font <<
      /R8 12 0 R
      /R10 13 0 R
      /R12 14 0 R
      /R14 15 0 R
      /R17 16 0 R
      /R20 17 0 R
      /R23 18 0 R
      /R27 19 0 R
    >>
    /ProcSet [ /PDF /Text ]
    /ExtGState <<
      /R7 20 0 R
    >>
  >>
>>
```
This does not prevent the PDF creator to put things on the page that (partly) go beyond this area.
This is legal and no error. Often happens with scanned pages.

You simply have to accept that and adjust your code accordingly if needed.
Default extraction flags in `page.get_text(...)` ignore anything not fully contained in the MediaBox. This can be changed by using adjusting the bit switches of these flags.

Method `page.get_image_info()` does not impose such restrictions and will report all images of the page.

**Reply by jmgomezsoriano** [reply]
Thanks for the answer.

Sorry, I can't share the example PDF because its information is confidential, and that document wasn't created by me. I'm not sure how my colleague create it, but I suspect that she generated an image jpeg from a PDF page, and create a PDF adding that image as a PDF page. I ignore what tools has used for that. I open the document with a PDF reader and I see the image with the same size of the page.  On Monday I'll ask her to try to replicate with other documents less confidential. The image inside the PDF is lightly bigger than the own page.

The print(doc.xref_object(page.xref)) prints this:

```python
<<
  /Contents [ 5 0 R 8 0 R 9 0 R 10 0 R ]
  /CropBox [ 0 0 595.27578 841.89 ]
  /MediaBox [ 0 0 595.27578 841.89 ]
  /Parent 2 0 R
  /Resources 6 0 R
  /Rotate 0
  /Type /Page
>>
```

The pdf has no text because it's only a image.
And the page.get_image_info() gives me the following:

```python
[
    {
         'number': 0, 
         'bbox': (-2.3249998776009306e-05, 0.0, 595.500244140625, 842.6346435546875),
         'transform': (595.500244140625, 0.0, -0.0, 842.6346435546875, -2.3249998776009306e-05, 0.0),
         'width': 1653,
         'height': 2339,
         'colorspace': 3,
         'cs-name': 'DeviceRGB',
         'xres': 96,
         'yres': 96,
         'bpc': 8,
         'size': 413746,
         'digest': b'N\xa8\x0eI\x887\xc84\xe9F\x17\x12"\xcdh\x14'
    }
]
```

However, I can't still understand because the image has a dimensions with page.rect, but when I try to draw a rectangle the starting point is different to (0, 0) and why I need to draw a bigger rectangle in the page than those defined in page.react.

**Reply by JorjMcKie** [reply]
> However, I can't still understand because the image has a dimensions with page.rect, but when I try to draw a rectangle the starting point is different to (0, 0) and why I need to draw a bigger rectangle in the page than those defined in page.react.

I don't see your problem.
Everything is ok with that page. It simply shows an image whose insertion area is larger / not fully inside than the MediaBox.
You seem to have a misconception: It regularly happens, that things exist outside the MediaBox - also text.
Some creator use this approach to store some hidden information there.
It also is not necessarily the "physical" page.

**Reply by jmgomezsoriano** [reply]
Thanks for your answer. I could reproduce the problem with pixelated documents. I have the following document:

[pixelated-image.pdf](https://github.com/pymupdf/PyMuPDF/files/14822488/pixelated-image.pdf)

And I execute the following code:

```python
doc = fitz.Document('pixelated-image.pdf', filetype="pdf")
page.draw_rect(page.rect, color=(0, 0, 1), width=2)
doc.save('pixelated-image-with-bounded-page.pdf')
doc.close()
```

And I get this document:

[pixelated-image-with-bounded-page.pdf](https://github.com/pymupdf/PyMuPDF/files/14822508/pixelated-image-with-bounded-page.pdf)

As you can see, the page.rect dimensions do not match with the real page size. And I don't know the reason.

**Reply by GitHubRulesOK** [reply]
Hmmm one of the common problems when working with PDF is using DPI so I wonder why there is 
```
         'xres': 96,
         'yres': 96,
```
It has no value for an image in a PDF thus commonly causes a misconception about scale which by default is 72 points per inch
NOT 72 DPI or 96 DPI but they are commonly seen as related to each other so a scale of 0.75 gets applied erroneously

Replace the image with 2 dots high by 2 dots wide now what is the resolution ? Images are not a "measure" of page it is  roughly 595x842  at 1x:1y units

[pixelated-image-with-bounded-paged.pdf](https://github.com/pymupdf/PyMuPDF/files/14828141/pixelated-image-with-bounded-paged.pdf)


So look inside the PDF with Notepad and you will see many objects such as scale set to 0.75 !! then a White Page scaled to 1.00 so needing units inflated to 133% 

```
stream
0.750000 0.000000 0.000000 -0.750000 0.000000 841.890015 cm
q
q
1.000000 0.000000 0.000000 1.000000 0.000000 0.000000 cm
1.000000 1.000000 1.000000 rg
0.000000 0.000000 m
793.700989 0.000000 l
793.700989 1122.520020 l
0.000000 1122.520020 l
h
f*
Q
```
Then the page box and Image but the secondary page overlay the BLUE box WITHOUT RESET is going to be affected by all those underlying rescalings.
```
8 0 obj
<</Length 44>>
stream

q
0 0 595.276 841.89 re
2 w
h
0 0 1 RG S
Q
```
so page size is 793.70101333 units x 1122.51999999 units = full page AT that time.

WHY does it happen from the page start? Is because Microsoft XPS likes to work in 75% scale so erroneously thinks A4 is 280 mm x 396 mm !!!!!

![image](https://github.com/pymupdf/PyMuPDF/assets/12754913/e91ce473-91e5-46d7-8538-5aa88e70c8a6)

**Reply by jmgomezsoriano** [reply]
Finally, I've resolved the problem converting the PDF to PDF before to draw the positions:

```python
doc = fitz.open(filename, fitz.Document(join('examples', 'doc.pdf'), filetype="pdf")).convert_to_pdf())
page = doc[0]
images = page.get_images(full=True)
image = images[0]
page.draw_rect(page.rect, color=(0, 0, 1), width=2)
```

The reason, I don't know, but it works.
